Sources/Coproduct.swift

```diff
@@ -0,0 +1,76 @@
+public protocol CoproductType {
+	associatedtype LeftType
+	associatedtype RightType
+
+	func fold<T>(onLeft: @escaping (LeftType) -> T, onRight: @escaping (RightType) -> T) -> T
+}
+
+public enum Coproduct<A,B>: CoproductType {
+	case left(A)
+	case right(B)
+
+	public func fold<T>(onLeft: @escaping (A) -> T, onRight: @escaping (B) -> T) -> T {
+		switch self {
+		case .left(let value):
+			return onLeft(value)
+		case .right(let value):
+			return onRight(value)
+		}
+	}
+}
+
+extension Coproduct where A: Equatable, B: Equatable {
+	public static func == (lhs: Coproduct, rhs: Coproduct) -> Bool {
+		return lhs.fold(
+			onLeft: { left in
+				rhs.fold(
+					onLeft: { left == $0 },
+					onRight: { _ in false })
+		},
+			onRight: { left in
+				rhs.fold(
```

Comment:
this should be called `right`, not `left`

Reply:
done

Reply:
done